metaflow_extensions/plugins/install_step_decorator.py

```diff
@@ -13,17 +13,88 @@
+    get_conda_python_executable,
+    get_pkg_spec,
     local_pip_install,
+    pip_install,
+    platform_arch_mismatch,
     running_in_local_env,
     up_to_project_root,
     upgrade_pip,
 )
 
 
+def prepare_step_wrapper(conda_step_prepare_step_environment: Callable) -> Callable:
+    """Appends commands after CondaStepDecorator._prepare_step_environment execution.
+
+    In the flow lifecycle this is executed prior to steps being explicitly executed:
+    in practice this occurs in conda_step_decorator.package_init, and it is only
+    executed on local steps (i.e. not batch steps). There is some duplication with
+    the prepare_step_wrapper method in the install_step_decorator module as the
+    bootstrap_commands method is executed immediately prior to the batch runtime,
+    and is never called locally.
+
+    In short, the difference between install_step_decorator.prepare_step_wrapper
+    and project_environment.bootstrap_wrapper is that:
+
+      * prepare_step_wrapper patches the local conda environment only (it is unable to
+        patch the batch conda environment because different of architectures, e.g.
+        MacOS on local vs Linux on batch)
+      * bootstrap_wrapper patches the batch conda env only, by definition.
+
+    Arguments:
+        conda_step_prepare_step_environment: CondaStepDecorator method to be wrapped
+
+    Returns:
+        The wrapped function
+    """
+
+    def wrapped_prepare_step_environment(self, *args, **kwargs):
+        from metaflow_extensions.config.metaflow_config import PREINSTALL_PKGS
+
+        env_id = conda_step_prepare_step_environment(self, *args, **kwargs)
+
+        # tl,dr; if this is a batch step, skip it, as package installation
+        #        for batch steps is done by bootstrap_wrapper, whereas
+        #        prepare_step_wrapper deals with local steps.
+        #
+        # Batch steps have a conda environment with an arch (Linux) that doesn't
+        # match the local arch (e.g. MacOS). In the "edge case" where the local
+        # arch and batch arch match (i.e. are both Linux) then no harm
+        # will be done by doubly installing the same packages, both here and
+        # with project_environment.bootstrap_wrapper
+        if platform_arch_mismatch(env_id):
+            return env_id
+
+        python_exec = get_conda_python_executable(env_id)
+        upgrade_pip(python_exec)  # Need newer features for install step
+
+        # Pre-install specified packages if the user has them
+        # installed in their local python env
+        for pkg_spec in filter(None, map(get_pkg_spec, PREINSTALL_PKGS)):
+            pip_install(python_exec, pkg_spec)
+
+        # # Install the local project if it conforms to the expected structure
+        # flow_path = Path(sys.argv[0])
+        # project_root = up_to_project_root(flow_path)
+        # if project_root:
+        #     pip_install(python_exec, project_root)
+        # return env_id
```

Comment:
Just went to test whether I could get things to break where `myproject` imports unavailable dependencies and in doing so realised that the return statement was accidentally commented out which borks things.

Reply:
This is also why the CI tests are failing.

And commenting the above logic as per my suggestion breaks everything because that logic (duplicated from `task_pre_step`) has to be run earlier because of `daps_utils` weird backwards dependency on `myproject`

Reply:
(sorry should have said that I'm still investigating, I've got half a report together on the full execution sequence to disentangle things)

Reply:
I would say that although the backwards dependency of `daps_utils` on `myproject` is weird, it's not particularly weird to expect the current project to be available at runtime, i.e. what `daps_utils` expects is that the following would be possible - which it currently isn't in with `environment=conda`:

```python
import myproject
class MyFlow(FlowSpec):
    ...
```

the duplication of the snippet from `task_pre_step` basically makes this possible, which I think is pretty reasonable isn't it?

Reply:
> which I think is pretty reasonable isn't it?
I guess so, because `myproject` isn't just another dependency like you would pass to `@conda`.